src/sniffer.c

```diff
@@ -7132,6 +7191,308 @@ int ssl_PollSniffer(WOLF_EVENT** events, int maxEvents, WOLF_EVENT_FLAG flags,
 }
 #endif
 
+
+#if defined(WOLFSSL_SNIFFER_KEYLOGFILE)
+
+#define CLIENT_RANDOM_LABEL_LENGTH 13
+#define CLIENT_RANDOM_LENGTH 32
+#define MASTER_SECRET_LENGTH 48
+#define CLIENT_RANDOM_BITS ((CLIENT_RANDOM_LENGTH) * 8)
+
+
+typedef struct SecretNode {
+    unsigned char clientRandom[CLIENT_RANDOM_LENGTH];
+    unsigned char masterSecret[MASTER_SECRET_LENGTH];
+    struct SecretNode* next;
+} SecretNode;
+
+
+/* Default to the same size hash table as the session table, but allow user to override */
+#ifndef WOLFSSL_SNIFFER_KEYLOGFILE_HASH_TABLE_SIZE
+#define WOLFSSL_SNIFFER_KEYLOGFILE_HASH_TABLE_SIZE HASH_SIZE
+#endif
+
+static THREAD_LS_T WOLFSSL_GLOBAL SecretNode* secretHashTable[WOLFSSL_SNIFFER_KEYLOGFILE_HASH_TABLE_SIZE] = {NULL};;
+#ifndef HAVE_C___ATOMIC
+static WOLFSSL_GLOBAL wolfSSL_Mutex secretListMutex;
+#endif
+
+
+static unsigned int secretHashFunction(unsigned char* clientRandom);
+
+#ifdef HAVE_C___ATOMIC
+    #define LOCK_SECRET_LIST() WC_DO_NOTHING
+    #define UNLOCK_SECRET_LIST() WC_DO_NOTHING
+#else
+    #define LOCK_SECRET_LIST() wc_LockMutex(&secretListMutex)
+    #define UNLOCK_SECRET_LIST() wc_UnLockMutex(&secretListMutex)
+#endif
+
+/*
+ * Basic polynomial hash function that maps a 32-byte client random value to an array index
+ */
+static unsigned int secretHashFunction(unsigned char* clientRandom)
+{
+    unsigned int hash = 0;
+    for (int i = 0; i < CLIENT_RANDOM_LENGTH; i++) {
+        hash = (hash * CLIENT_RANDOM_BITS + clientRandom[i]) % WOLFSSL_SNIFFER_KEYLOGFILE_HASH_TABLE_SIZE;
+    }
+    return hash;
+}
+
+static int addSecretNode(unsigned char* clientRandom, unsigned char* masterSecret, char* error)
+{
+    SecretNode* newSecretNode = (SecretNode*)XMALLOC(sizeof(SecretNode), NULL, DYNAMIC_TYPE_SNIFFER_KEYLOG_NODE);
+    if (newSecretNode == NULL) {
+        SetError(MEMORY_STR, error, NULL, 0);
+        return WOLFSSL_SNIFFER_ERROR;
+    }
+
+    XMEMCPY(newSecretNode->clientRandom, clientRandom, CLIENT_RANDOM_LENGTH);
+    XMEMCPY(newSecretNode->masterSecret, masterSecret, MASTER_SECRET_LENGTH);
+
+    LOCK_SECRET_LIST();
+
+    unsigned int index = secretHashFunction(clientRandom);
+    newSecretNode->next = NULL;
+
+    if (secretHashTable[index] == NULL) {
+        secretHashTable[index] = newSecretNode;
+    }
+    else {
+        SecretNode* current = secretHashTable[index];
+        while (current != NULL) {
+            if (memcmp(current->clientRandom, clientRandom, CLIENT_RANDOM_LENGTH) == 0) {
+                // BRN-sniffer-TODO: what if the same client random has a different master secret?
```

Comment:
Are you saying there are multiple client random entries? Use the first available based on order in file. Don't consider that an error. I could see this happening for WOLFSSL_GENSEED_FORTEST cases...